## Telco Customer Churn Prediction

### Data Exploration Notebook

In [1]:
# Importing necessary libraries
import pandas as pd

In [2]:
# Adjusting visibility
pd.set_option("display.max_column", None)

In [ ]:
# Importing data
df = pd.read_csv('data/WA_Fn-UseC_-Telco-Customer-Churn.csv')

In [4]:
# We check the first rows of data
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  No phone service             DSL             No          Yes   
1                No             DSL            Yes           No   
2                No             DSL            Yes          Yes   
3  No phone service             DSL            Yes           No   
4                No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  \
0              Yes           Electronic check           29.85        29.85   
1               No               Mailed check           56.95       1889.5   
2              Yes               Mailed check           53.85       108.15   
3               No  Bank transfer (automatic)           42.30      1840.75   
4              Yes           Electronic check           70.70       151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes

In [5]:
df.describe()

SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000

We can notice that despite TotalCharges is a numerical variable it doesn't appear when using `describe()` which might be because of NaN or Null values.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


TotalCharges isn't being considered a numerical variable in the dataframe, we convert it to a numerical type to check the values causing the problem

In [7]:
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors="coerce")

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


Now we can see 11 null values in TotalCharges

In [9]:
df[df['TotalCharges'].isna()]

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
488   4472-LVYGI  Female              0     Yes        Yes       0   
753   3115-CZMZD    Male              0      No        Yes       0   
936   5709-LVOEQ  Female              0     Yes        Yes       0   
1082  4367-NUYAO    Male              0     Yes        Yes       0   
1340  1371-DWPAZ  Female              0     Yes        Yes       0   
3331  7644-OMVMY    Male              0     Yes        Yes       0   
3826  3213-VVOLG    Male              0     Yes        Yes       0   
4380  2520-SGTTA  Female              0     Yes        Yes       0   
5218  2923-ARZLG    Male              0     Yes        Yes       0   
6670  4075-WKNIU  Female              0     Yes        Yes       0   
6754  2775-SEFEE    Male              0      No        Yes       0   

     PhoneService     MultipleLines InternetService       OnlineSecurity  \
488            No  No phone service             DSL                  Yes   
753           Yes                No              No  No internet service   
936           Yes                No             DSL                  Yes   
1082          Yes               Yes              No  No internet service   
1340           No  No phone service             DSL                  Yes   
3331          Yes                No              No  No internet service   
3826          Yes               Yes              No  No internet service   
4380          Yes                No              No  No internet service   
5218          Yes                No              No  No internet service   
6670          Yes               Yes             DSL                   No   
6754          Yes               Yes             DSL                  Yes   

             OnlineBackup     DeviceProtection          TechSupport  \
488                    No                  Yes                  Yes   
753   No internet service  No internet service  No internet service   
936                   Yes                  Yes                   No   
1082  No internet service  No internet service  No internet service   
1340                  Yes                  Yes                  Yes   
3331  No internet service  No internet service  No internet service   
3826  No internet service  No internet service  No internet service   
4380  No internet service  No internet service  No internet service   
5218  No internet service  No internet service  No internet service   
6670                  Yes                  Yes                  Yes   
6754                  Yes                   No                  Yes   

              StreamingTV      StreamingMovies  Contract PaperlessBilling  \
488                   Yes                   No  Two year              Yes   
753   No internet service  No internet service  Two year               No   
936                   Yes                  Yes  Two year               No   
1082  No internet service  No internet service  Two year               No   
1340                  Yes                   No  Two year               No   
3331  No internet service  No internet service  Two year               No   
3826  No internet service  No internet service  Two year               No   
4380  No internet service  No internet service  Two year               No   
5218  No internet service  No internet service  One year              Yes   
6670                  Yes                   No  Two year               No   
6754                   No                   No  Two year              Yes   

                  PaymentMethod  MonthlyCharges  TotalCharges Churn  
488   Bank transfer (automatic)           52.55           NaN    No  
753                Mailed check           20.25           NaN    No  
936                Mailed check           80.85           NaN    No  
1082               Mailed check           25.75           NaN    No  
1340    Credit card (automatic)           56.05           NaN    No  
3331               Mailed check           19.85       

This is mainly caused by the 0 tenure, since the customers didn't complete a full month yet<br>
So we fill with 0 for NaN TotalCharges

In [12]:
df['TotalCharges'] = df['TotalCharges'].fillna(0)

In [13]:
df.isna().sum().sum()

np.int64(0)

In [14]:
# CustomerID won't be necessary for the analysis so we drop it
df.drop("customerID", axis=1, inplace=True)

In [17]:
# Now we need to encode the categorical variables, we check the distribution of each column
grouped = ['gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity',
           'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
           'PaymentMethod', 'Churn']

for grp in grouped:
    print(df.groupby(grp).size())
    print('------')

gender
Female    3488
Male      3555
dtype: int64
------
Partner
No     3641
Yes    3402
dtype: int64
------
Dependents
No     4933
Yes    2110
dtype: int64
------
PhoneService
No      682
Yes    6361
dtype: int64
------
MultipleLines
No                  3390
No phone service     682
Yes                 2971
dtype: int64
------
InternetService
DSL            2421
Fiber optic    3096
No             1526
dtype: int64
------
OnlineSecurity
No                     3498
No internet service    1526
Yes                    2019
dtype: int64
------
OnlineBackup
No                     3088
No internet service    1526
Yes                    2429
dtype: int64
------
DeviceProtection
No                     3095
No internet service    1526
Yes                    2422
dtype: int64
------
TechSupport
No                     3473
No internet service    1526
Yes                    2044
dtype: int64
------
StreamingTV
No                     2810
No internet service    1526
Yes                    2707
dtype

We see that most of the variables are binary (Yes/No), however some variables have more.
- For `MultipleLines` we can encode No phone service as 0, No as 1 since the user has 1 phone and Yes as 2 for having multiple phones
- The number of customers having no internet service is the same in various variables (1526) so we can encode the variables that have it as a value in a similar way (`OnlineSecurity`, `OnlineBackup`, `DeviceProtection`, `TechSupport`, `StreamingTV`, `StreamingMovies`) setting Yes to 1 No internet service and No to 0
- `Contract` can be encoded ordinally setting `Month-to-month` to 0, `One year` to 1 and `Two year` to 2
- `PaymentMethod` can be split to dummies, we can also engineer a new variable from it, to check if manual/automatic payment has an effect on churn

In [18]:
df['HasInternet'] = ~df['OnlineBackup'].isin(['No internet service'])

In [19]:
# Encoding gender
df['gender'] = df['gender'].map({'Male' : 1, 'Female' : 0})

In [20]:
# we encode binary variables that have "yes/no" as values
yes_no_vars = ['Partner', 'Dependents', 'PhoneService', 'PaperlessBilling', 'Churn']

for yn in yes_no_vars:
    df[yn] = df[yn].map({'Yes' : 1, 'No' : 0})

In [21]:
# Encoding the number of phone lines and contract
df['MultipleLines'] = df['MultipleLines'].map({'No phone service' : 0, 'No' : 1, 'Yes' : 2})
df['Contract'] = df['Contract'].map({'Month-to-month' : 0, 'One year' : 1, 'Two year' : 2})

In [22]:
internet_vars = ['StreamingMovies', 'StreamingTV', 'TechSupport', 'DeviceProtection', 'OnlineSecurity', 'OnlineBackup']
for i in internet_vars:
    df[i] = df[i].map({'No internet service' : 0, 'No' : 0, 'Yes' : 1})

In [23]:
# Engineering the automatic_pay variable
df['automatic_pay'] = df['PaymentMethod'].isin(['Bank transfer (automatic)','Credit card (automatic)'])

In [24]:
# we get the dummy variables
df = pd.get_dummies(df, columns=['InternetService','PaymentMethod'], prefix='', prefix_sep='')

In [25]:
df.head()

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0       0              0        1           0       1             0   
1       1              0        0           0      34             1   
2       1              0        0           0       2             1   
3       1              0        0           0      45             0   
4       0              0        0           0       2             1   

   MultipleLines  OnlineSecurity  OnlineBackup  DeviceProtection  TechSupport  \
0              0               0             1                 0            0   
1              1               1             0                 1            0   
2              1               1             1                 0            0   
3              0               1             0                 1            1   
4              1               0             0                 0            0   

   StreamingTV  StreamingMovies  Contract  PaperlessBilling  MonthlyCharges  \
0            0                0         0                 1           29.85   
1            0                0         1                 0           56.95   
2            0                0         0                 1           53.85   
3            0                0         1                 0           42.30   
4            0                0         0                 1           70.70   

   TotalCharges  Churn  HasInternet  automatic_pay    DSL  Fiber optic     No  \
0         29.85      0         True          False   True        False  False   
1       1889.50      0         True          False   True        False  False   
2        108.15      1         True          False   True        False  False   
3       1840.75      0         True           True   True        False  False   
4        151.65      1         True          False  False         True  False   

   Bank transfer (automatic)  Credit card (automatic)  Electronic check  \
0                      False                    False              True   
1                      False                    False             False   
2                      False                    False             False   
3                       True                    False             False   
4                      False                    False              True   

   Mailed check  
0         False  
1          True  
2          True  
3         False  
4         False

In [ ]:
# to avoid the dummies linearity trap
df.drop(['No','Mailed check'], axis=1, inplace=True)

In [27]:
df.head()

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0       0              0        1           0       1             0   
1       1              0        0           0      34             1   
2       1              0        0           0       2             1   
3       1              0        0           0      45             0   
4       0              0        0           0       2             1   

   MultipleLines  OnlineSecurity  OnlineBackup  DeviceProtection  TechSupport  \
0              0               0             1                 0            0   
1              1               1             0                 1            0   
2              1               1             1                 0            0   
3              0               1             0                 1            1   
4              1               0             0                 0            0   

   StreamingTV  StreamingMovies  Contract  PaperlessBilling  MonthlyCharges  \
0            0                0         0                 1           29.85   
1            0                0         1                 0           56.95   
2            0                0         0                 1           53.85   
3            0                0         1                 0           42.30   
4            0                0         0                 1           70.70   

   TotalCharges  Churn  HasInternet  automatic_pay    DSL  Fiber optic  \
0         29.85      0         True          False   True        False   
1       1889.50      0         True          False   True        False   
2        108.15      1         True          False   True        False   
3       1840.75      0         True           True   True        False   
4        151.65      1         True          False  False         True   

   Bank transfer (automatic)  Credit card (automatic)  Electronic check  
0                      False                    False              True  
1                      False                    False             False  
2                      False                    False             False  
3                       True                    False             False  
4                      False                    False              True

In [28]:
# Encoding True/False variables
true_false_vars = ['DSL','Fiber optic','Bank transfer (automatic)','Credit card (automatic)','Electronic check',
                   'automatic_pay','HasInternet']
for tf in true_false_vars:
    df[tf] = df[tf].map({True : 1, False : 0})

In [29]:
df.head()

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0       0              0        1           0       1             0   
1       1              0        0           0      34             1   
2       1              0        0           0       2             1   
3       1              0        0           0      45             0   
4       0              0        0           0       2             1   

   MultipleLines  OnlineSecurity  OnlineBackup  DeviceProtection  TechSupport  \
0              0               0             1                 0            0   
1              1               1             0                 1            0   
2              1               1             1                 0            0   
3              0               1             0                 1            1   
4              1               0             0                 0            0   

   StreamingTV  StreamingMovies  Contract  PaperlessBilling  MonthlyCharges  \
0            0                0         0                 1           29.85   
1            0                0         1                 0           56.95   
2            0                0         0                 1           53.85   
3            0                0         1                 0           42.30   
4            0                0         0                 1           70.70   

   TotalCharges  Churn  HasInternet  automatic_pay  DSL  Fiber optic  \
0         29.85      0            1              0    1            0   
1       1889.50      0            1              0    1            0   
2        108.15      1            1              0    1            0   
3       1840.75      0            1              1    1            0   
4        151.65      1            1              0    0            1   

   Bank transfer (automatic)  Credit card (automatic)  Electronic check  
0                          0                        0                 1  
1                          0                        0                 0  
2                          0                        0                 0  
3                          1                        0                 0  
4                          0                        0                 1

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 25 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   gender                     7043 non-null   int64  
 1   SeniorCitizen              7043 non-null   int64  
 2   Partner                    7043 non-null   int64  
 3   Dependents                 7043 non-null   int64  
 4   tenure                     7043 non-null   int64  
 5   PhoneService               7043 non-null   int64  
 6   MultipleLines              7043 non-null   int64  
 7   OnlineSecurity             7043 non-null   int64  
 8   OnlineBackup               7043 non-null   int64  
 9   DeviceProtection           7043 non-null   int64  
 10  TechSupport                7043 non-null   int64  
 11  StreamingTV                7043 non-null   int64  
 12  StreamingMovies            7043 non-null   int64  
 13  Contract                   7043 non-null   int64

In [31]:
df.isna().sum().sum()

np.int64(0)

Next we can normalize the numerical variables (`tenure`, `MonthlyCharges`, `TotalCharges`) or leave it as it is

In [33]:
to_scale = ['tenure','MonthlyCharges','TotalCharges']

In [34]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()

df[to_scale] = scaler.fit_transform(df[to_scale])

In [35]:
df.head()

gender  SeniorCitizen  Partner  Dependents    tenure  PhoneService  \
0       0              0        1           0  0.013889             0   
1       1              0        0           0  0.472222             1   
2       1              0        0           0  0.027778             1   
3       1              0        0           0  0.625000             0   
4       0              0        0           0  0.027778             1   

   MultipleLines  OnlineSecurity  OnlineBackup  DeviceProtection  TechSupport  \
0              0               0             1                 0            0   
1              1               1             0                 1            0   
2              1               1             1                 0            0   
3              0               1             0                 1            1   
4              1               0             0                 0            0   

   StreamingTV  StreamingMovies  Contract  PaperlessBilling  MonthlyCharges  \
0            0                0         0                 1        0.115423   
1            0                0         1                 0        0.385075   
2            0                0         0                 1        0.354229   
3            0                0         1                 0        0.239303   
4            0                0         0                 1        0.521891   

   TotalCharges  Churn  HasInternet  automatic_pay  DSL  Fiber optic  \
0      0.003437      0            1              0    1            0   
1      0.217564      0            1              0    1            0   
2      0.012453      1            1              0    1            0   
3      0.211951      0            1              1    1            0   
4      0.017462      1            1              0    0            1   

   Bank transfer (automatic)  Credit card (automatic)  Electronic check  
0                          0                        0                 1  
1                          0                        0                 0  
2                          0                        0                 0  
3                          1                        0                 0  
4                          0                        0                 1

In [36]:
df.dtypes

gender                         int64
SeniorCitizen                  int64
Partner                        int64
Dependents                     int64
tenure                       float64
PhoneService                   int64
MultipleLines                  int64
OnlineSecurity                 int64
OnlineBackup                   int64
DeviceProtection               int64
TechSupport                    int64
StreamingTV                    int64
StreamingMovies                int64
Contract                       int64
PaperlessBilling               int64
MonthlyCharges               float64
TotalCharges                 float64
Churn                          int64
HasInternet                    int64
automatic_pay                  int64
DSL                            int64
Fiber optic                    int64
Bank transfer (automatic)      int64
Credit card (automatic)        int64
Electronic check               int64
dtype: object

The preprocessing is done now, next step would be training the model.